### PR TITLE
Add roadmap to kueue

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind document

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/issues/424

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,6 +84,21 @@ to get document access.
 - [bit.ly/kueue-controller-design](https://bit.ly/kueue-controller-design)
 presents the detailed design of the controller.
 
+## Roadmap
```
Comment (alculquicondor) [inline README.md:87]: In general, add links to issues if they exist.

**Comment by alculquicondor** [comment]
I think that, if we are going to publish a roadmap, we should have some priorities and estimated times. Otherwise it might bring more questions from readers than providing answers.

Maybe we can discuss this in the WG Batch meeting?

**Comment by alculquicondor** [comment]
The agenda for Dec 22nd is still empty https://docs.google.com/document/d/1XOeUN-K0aKmJJNq7H07r74n-mGgSFyiEDQ3ecwsGhec/edit?pli=1#

**Comment by kerthcet** [comment]
> I think that, if we are going to publish a roadmap, we should have some priorities and estimated times. Otherwise it might bring more questions from readers than providing answers.
> 
> Maybe we can discuss this in the WG Batch meeting?

At the least, we can tell people how we will influence the project direction. But it will be really great if we can have a developing plan based on priorities. A roadmap for full 2023?

I think we can talk about this in the community meeting to get some feedbacks what users want most. I'll add to agenda later.

**Comment by ahg-g** [comment]
Thanks for documenting the roadmap, I will take a look after you address the open comments from aldo.

File: README.md (hunk 1 above)
Comment (kerthcet) [reply README.md:87]: Done.

**Comment by alculquicondor** [comment]
looks good, but should we include time estimates to avoid questions?

**Comment by kerthcet** [comment]
> looks good, but should we include time estimates to avoid questions?

update this as roadmap for 2023, make sense? We can discuss this later in community meeting.

File: README.md [hunk 2]
```diff
@@ -81,6 +84,26 @@ to get document access.
...
+
+This is a high-level overview of the main priorities for Kueue, in expected order of release:
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83)
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply README.md]: Added the plan

File: README.md [hunk 3]
```diff
@@ -81,6 +84,26 @@ to get document access.
...
 presents the detailed design of the controller.
 
+## Roadmap
+
+This is a high-level overview of the main priorities for 2023, in expected order of release:
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
+- More advanced allocation strategies for better resource sharing [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
+- Integration with cluster-autoscaler for guaranteed resource provisioning
+- Integration with common custom workloads [#74](https://github.com/kubernetes-sigs/kueue/issues/74):
```
Comment (denkensk) [inline README.md:94]: Why put this in the last place? I think this should be a higher priority.

Comment (alculquicondor) [reply README.md:94]: This is ordered according to our current involvement/capacity. We certainly want this, but there isn't anybody working on it at the moment.

Comment (denkensk) [reply README.md:94]: Okay, got it. Thank you for explaining this, Aldo.

File: README.md [hunk 4]
```diff
@@ -81,6 +84,26 @@ to get document access.
...
+  - Spark
+  - Workflows (Tekton, Argo, etc.)
+
+### Long-Term Features
+
+These are features that we aim to have in the long-term, in no particular order:
+
+- Multi-cluster support
+- Dashboard for management and monitoring for administrators
+- Budget support
```
Comment (alculquicondor) [inline README.md:105]: maybe on these two it's more realistic to make progress in 2023, as opposed to multi-cluster

Comment (alculquicondor) [reply README.md:105]: I don't think the roadmap needs to contain every possible feature. Just those that are bigger effort.

Comment (denkensk) [reply README.md:105]: Maybe you can add #28 behind budget support, then if someone want to learn details or background, they can read or ask in that issue.

Comment (kerthcet) [reply README.md:105]: updated, actually there's no particular order, just a rough plan.

File: README.md [hunk 5]
```diff
@@ -81,6 +84,26 @@ to get document access.
 - [bit.ly/kueue-controller-design](https://bit.ly/kueue-controller-design)
 presents the detailed design of the controller.
 
+## Roadmap
+
+This is a high-level overview of the main priorities for 2023, in expected order of release:
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
```
Comment (alculquicondor) [inline README.md:91]: Let's add cooperative preemption as a separate item #477

File: README.md [hunk 6]
```diff
@@ -81,6 +84,26 @@ to get document access.
...
+- More advanced allocation strategies for better resource sharing [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
+- Integration with cluster-autoscaler for guaranteed resource provisioning
+- Integration with common custom workloads [#74](https://github.com/kubernetes-sigs/kueue/issues/74):
+  - Kubeflow (TFJob, MPIJob, etc.)
+  - Spark
```
Comment (alculquicondor) [inline README.md:96]: Let's add Ray.io

File: README.md [hunk 7]
```diff
@@ -81,6 +84,26 @@ to get document access.
 - [bit.ly/kueue-controller-design](https://bit.ly/kueue-controller-design)
 presents the detailed design of the controller.
 
+## Roadmap
+
+This is a high-level overview of the main priorities for 2023, in expected order of release:
```
Comment (alculquicondor) [inline README.md:89]: Maybe for a separate PR, but we should have a summary of existing features (a more extended version of what's mentioned in line 16).

File: README.md [hunk 8]
```diff
@@ -81,6 +84,26 @@ to get document access.
...
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
+- More advanced allocation strategies for better resource sharing [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
+- Integration with cluster-autoscaler for guaranteed resource provisioning
+- Integration with common custom workloads [#74](https://github.com/kubernetes-sigs/kueue/issues/74):
+  - Kubeflow (TFJob, MPIJob, etc.)
+  - Spark
+  - Workflows (Tekton, Argo, etc.)
+
+### Long-Term Features
```
Comment (alculquicondor) [inline README.md:99]: @ahg-g can you think of a better heading here?

File: README.md (hunk 7 above)
Comment (kerthcet) [reply README.md:89]: This worths a separate PR. Maybe together with v0.3.0.

Comment (alculquicondor) [reply README.md:89]: Yes :)

File: README.md [hunk 9]
```diff
@@ -89,15 +89,15 @@ presents the detailed design of the controller.
 This is a high-level overview of the main priorities for 2023, in expected order of release:
 
 - Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
+- Cooperative preemption support with checkpoint mechanism [#477](https://github.com/kubernetes-sigs/kueue/issues/477)
```
Comment (alculquicondor) [inline README.md:92]:
```suggestion
- Cooperative preemption support for workloads that implement checkpointing [#477](https://github.com/kubernetes-sigs/kueue/issues/477)
```

we won't be providing the checkpoint mechanism.

Comment (kerthcet) [reply README.md]: sg.

**Review by alculquicondor — COMMENTED** [review]
/approve

@ahg-g anything to add?

File: README.md [hunk 10]
```diff
@@ -81,6 +84,26 @@ to get document access.
 - [bit.ly/kueue-controller-design](https://bit.ly/kueue-controller-design)
 presents the detailed design of the controller.
 
+## Roadmap
+
+This is a high-level overview of the main priorities for 2023, in expected order of release:
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
+- Cooperative preemption support for workloads that implement checkpointing [#477](https://github.com/kubernetes-sigs/kueue/issues/477)
+- More advanced allocation strategies for better resource sharing [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
```
Comment (ahg-g) [inline README.md:93]: This is a bit vague, perhaps specifically say: "Tunable flavor assignment to tradeoff cost vs borrowing"

Comment (kerthcet) [reply README.md]: Updated, I thought this was merged already. 🙃

**Comment by kerthcet** [comment]
I think we should also add a page to the website next.

File: README.md [hunk 11]
```diff
@@ -81,6 +84,26 @@ to get document access.
 - [bit.ly/kueue-controller-design](https://bit.ly/kueue-controller-design)
 presents the detailed design of the controller.
 
+## Roadmap
+
+This is a high-level overview of the main priorities for 2023, in expected order of release:
+
+- Job preemption to reclaim borrowed quota and to accommodate high priority jobs [#83](https://github.com/kubernetes-sigs/kueue/issues/83), this is planned for v0.3
+- Cooperative preemption support for workloads that implement checkpointing [#477](https://github.com/kubernetes-sigs/kueue/issues/477)
+- More flavor assignment strategies support, e.g. _minimizing cost_ vs _minimizing borrowing_ [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
```
Comment (ahg-g) [inline README.md:93]:
```suggestion
- Flavor assignment strategies support, e.g. _minimizing cost_ vs _minimizing borrowing_ [#312](https://github.com/kubernetes-sigs/kueue/issues/312)
```

Comment (kerthcet) [reply README.md:93]: updated

**Review by alculquicondor — COMMENTED** [review]
/approve

**Pull request merged**